storage/postgres/metadata.go

```diff
@@ -68,40 +70,124 @@
 	return meta, nil
 }
 
-func (s *PGStore) SaveMeta(ty string, id string, m core.Metadata) error {
+func (s *PGStore) SaveMeta(id, timestamp, targetType, targetID, key, value string) error {
 	tx, _ := s.Conn().Begin(context.TODO())
 
-	for key, value := range m {
-		ib := sqlbuilder.NewInsertBuilder()
-		ib.InsertInto(s.table("metadata"))
-		ib.Cols(
-			"meta_target_type",
-			"meta_target_id",
-			"meta_key",
-			"meta_value",
-		)
-		ib.Values(ty, id, key, string(value))
+	ib := sqlbuilder.NewInsertBuilder()
+	ib.InsertInto(s.table("metadata"))
+	ib.Cols(
+		"meta_id",
+		"meta_target_type",
+		"meta_target_id",
+		"meta_key",
+		"meta_value",
+		"timestamp",
+	)
+	ib.Values(id, targetType, targetID, key, string(value), timestamp)
 
-		sqlq, args := ib.BuildWithFlavor(sqlbuilder.PostgreSQL)
+	sqlq, args := ib.BuildWithFlavor(sqlbuilder.PostgreSQL)
 
-		_, err := tx.Exec(
-			context.TODO(),
-			sqlq,
-			args...,
-		)
+	_, err := tx.Exec(
+		context.TODO(),
+		sqlq,
+		args...,
+	)
 
-		if err != nil {
-			tx.Rollback(context.TODO())
+	if err != nil {
+		tx.Rollback(context.TODO())
 
-			return err
-		}
+		return err
 	}
 
-	err := tx.Commit(context.TODO())
+	err = tx.Commit(context.TODO())
 
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
+
+func (s *PGStore) FindMeta(q query.Query) (query.Cursor, error) {
+	q.Limit = int(math.Max(-1, math.Min(float64(q.Limit), 100))) + 1
+
+	c := query.Cursor{}
+
+	in := sqlbuilder.NewSelectBuilder()
+	in.Select(
+		"meta_id",
+		"meta_target_type",
+		"meta_target_id",
+		"meta_key",
+		"meta_value",
+	)
+	in.From("metadata")
+	in.OrderBy("meta_id desc")
+	in.Limit(q.Limit)
+
+	sqlq, args := in.BuildWithFlavor(sqlbuilder.PostgreSQL)
+	if viper.GetBool("debug") {
+		fmt.Println(sqlq, args)
+	}
+
+	rows, err := s.Conn().Query(
+		context.TODO(),
+		sqlq,
+		args...,
+	)
+
+	if err != nil {
+		return c, err
+	}
+
+	foundMetadata := map[int64]core.Metadata{}
+
+	for rows.Next() {
+
+		md := core.Metadata{}
+
+		var (
```

Comment:
There were some issues with this implementation:

1. You initialize results with against the length of an empty map, so it is an empty array. This will always cause a panic when you do `results[md.metaID] = md` later.
2. Metadata does not have any of these fields you reference when scanning the row. The type is a wrapper around a `map[string]json.RawMessage`, that's why they are unmarshalled into standalone vars

Although, I do agree I can avoid iterating through my map and just initialize a slice that is of length `q.Limit`. I initially implemented this and then decided to just remove this function from both storage implementations and the interface since I'm no longer using it to determine the last metadata id (originally I had been in order to follow the pattern set for finding the last transaction), and it seems out of scope to include in this PR without calling it in the business logic of the ledger